src/shims/io.rs

```diff
@@ -0,0 +1,179 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+
+use crate::stacked_borrows::Tag;
+use crate::*;
+
+#[derive(Default)]
+pub struct FileHandler {
+    files: HashMap<i32, File>,
+    flags: HashMap<i32, i32>,
+    low: i32,
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn open(
+        &mut self,
+        path_op: OpTy<'tcx, Tag>,
+        flag_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let flag = this.read_scalar(flag_op)?.to_i32()?;
+
+        if flag
+            != this
+                .eval_path_scalar(&["libc", "O_RDONLY"])?
+                .unwrap()
+                .to_i32()?
+            && flag
+                != this
+                    .eval_path_scalar(&["libc", "O_CLOEXEC"])?
+                    .unwrap()
+                    .to_i32()?
+        {
+            throw_unsup_format!("Unsupported flag {:#x}", flag);
+        }
+
+        let path_bytes = this
+            .memory()
+            .read_c_str(this.read_scalar(path_op)?.not_undef()?)?;
+        let path = std::str::from_utf8(path_bytes)
+            .map_err(|_| err_unsup_format!("{:?} is not a valid utf-8 string", path_bytes))?;
+
+        match File::open(path) {
+            Ok(file) => {
+                let mut fh = &mut this.machine.file_handler;
+                fh.low += 1;
+                fh.files.insert(fh.low, file);
+                fh.flags.insert(fh.low, flag);
+                Ok(fh.low)
+            }
+
+            Err(e) => {
+                this.machine.last_error = e.raw_os_error().unwrap() as u32;
+                Ok(-1)
+            }
+        }
+    }
+
+    fn fcntl(
+        &mut self,
+        fd_op: OpTy<'tcx, Tag>,
+        cmd_op: OpTy<'tcx, Tag>,
+        arg_op: Option<OpTy<'tcx, Tag>>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+        let cmd = this.read_scalar(cmd_op)?.to_i32()?;
+
+        if cmd
+            == this
+                .eval_path_scalar(&["libc", "F_SETFD"])?
+                .unwrap()
+                .to_i32()?
+        {
+            let flag = this.read_scalar(arg_op.unwrap())?.to_i32()?;
+            this.machine.file_handler.flags.insert(fd, flag);
```

Comment:
How should I handle the flag changes for a file that is already opened? What should happen if someone decides to enable the `O_CREAT` flag for a file using `fcntl`?

Reply:
idk, I'd be fine with erroring as unsupported until someone hits a use-case. Or do whatever the libc docs state

Reply:
Well that's the thing, when someone implements  `execve`  they would need to take care of this flag but at the moment nothing uses `execve`.